### ISSUE TITLE
[Feature Request] Cardinality constraint

### ISSUE THREAD

**Issue by swierh** [pytorch/botorch#1749]
# 🚀 Feature Request

We've got a problem where we have a constraint on the number of non-zero inputs. Out of 10 possible features, we can only select 5 at a time for any sample in the q-batch. This is because in a certain sample we can only use a maximum of 5 different ingredients which we have to pick from 10 possible ingredients.

We tried implementing this by using a penalized acquisition function by adding a loss on the cardinality. This works, but a constraint seems like a better place for this.

Also, when we add a normal inequality constraint (e.g. x1 + x2 > 0), the optimizer doesn't seem to find solutions with low cardinality.

We've experimented with adding the cardinality constraint as a non-linear constraint, but it appears that have to write a custom sampler to get this working.

Is there a cleaner way of doing this, what would you suggest as the right approach for this?

All help will be much appreciated!

The code we've implemented for the penalty is as follows:

```python
class CardinalityPenalty(torch.nn.Module):
    r"""Cardinality penalty class to be added to any arbitrary acquisition function
    to construct a PenalizedAcquisitionFunction."""

    def __init__(self, init_point: Tensor, cardinality: float, epsilon: float = 0.01):
        r"""Initializing cardinality regularization.

        Args:
            init_point: The "1 x dim" reference point against which
                we want to regularize.
            cardinality: The cardinality below which penality is 0.
        """
        super().__init__()
        self.init_point = init_point
        self.allowed_cardinality = cardinality
        self.epsilon = epsilon

    def forward(self, X: Tensor) -> Tensor:
        r"""
        Args:
            X: A "batch_shape x q x dim" representing the points to be evaluated.
            The values in X must be -1 ≤ x ≤ 1.

        Returns:
            A tensor of size "batch_shape" representing the acqfn for each q-batch.
        """

        # We allow for a small deviation from 0, and we don't care about the sign:
        X_rel = (X - self.init_point).abs().sub(self.epsilon).clip(0)

        cardinality = torch.norm(X_rel, p=0, dim=-1)

        # If the cardinality is too high, we want the point to move to the closest
        # allowed plane, so we penalize the smallest non-zero value:
        min_non_zero = (X_rel + (X_rel == 0)).min(dim=-1).values

        # We penalize the cardinality if it's higher than the allowed cardinality.
        # At or below that cardinality, we don't care about the non-zero values,
        # which is why we subtract 1.
        # When the cardinality is higher than the allowed value, we start penalizing
        # the smallest non-zero value.
        penalty = (cardinality - self.allowed_cardinality - 1 + min_non_zero).clip(0)

        # The penalty must be a single value for the entire batch:
        penalty = penalty.mean(dim=-1) ** 2
        return penalty
```

And we then wrap our acquisition function as follows:

```python
acqf = qKnowledgeGradient(model, num_fantasies=10)
penalty = CardinalityPenalty(
    torch.tensor(np.zeros_like(X[0]), dtype=torch.float64, device=device),
    cardinality=5,
)
acqf = PenalizedAcquisitionFunction(acqf, penalty, 1.0)

x_batch, _ = optimize_acqf(acq_function=acqf, ...)
```

**Comment by eytan** [comment]
cc @dme65 , @bletham , @qingfeng10. 

Is the 5 at a time a hard constraint or some ideal number that you’d like to stick to? I wonder if some of the modeling+optimization approaches from  https://arxiv.org/pdf/2203.01900.pdf could be useful.

**Comment by jduerholt** [comment]
Hi, I call this kind of constraint an `NChooseK` constraint, because out of N ingredients you are able to pick k. In some discussion in the ax issues with @sgbaird ,it was shown that you can express this via a continuous relaxation as nonlinear inequality constraint, which is exactly the relaxation which is also used in the paper which @eytan mentioned above. I try to dig this discussion out later.

Currently I am implementing this in our own botorch wrapper focusing on materials discovery. Hopefully it is finished next week. In this method the NchooseK constraints are transformed into the botorch definition of a nonlinear inequality:

https://github.com/experimental-design/bofire/blob/cc33f21f4d33c3a58e20e88c8694a14c12a2713e/bofire/utils/torch_tools.py#L88

You can take this defintions and just give it to `optimize_acqf`.

Then you just need to setup initial conditions adhering to the constraints. This can be done by our Polytopesampler based on the polytopesampler of botorch which is also able to handle NChooseK constraints. 

Here is an example in which the NChooseK constraint is combined with an additional linear equality (mixture constraint)

```python
from bofire.data_models.constraints.api import (
    LinearInequalityConstraint,
    LinearEqualityConstraint,
    NChooseKConstraint,
)
from bofire.data_models.domain.api import Domain
from bofire.data_models.features.api import CategoricalInput, ContinuousInput

import bofire.strategies.api as strategies
import bofire.data_models.strategies.api as data_models
from bofire.utils.torch_tools import get_nchoosek_constraints

features = [ContinuousInput(key=f"if{i+1}", lower_bound=0, upper_bound=1) for i in range(10)]

constraints = [
    LinearEqualityConstraint(features=[f"if{i+1}" for i in range(10)], coefficients=[1.0]*10, rhs=1),
    NChooseKConstraint(features=[f"if{i+1}" for i in range(10)], min_count=0, max_count=5, none_also_valid=True)
    ]


domain = Domain(
        input_features=features,
        constraints=constraints,
    )

data_model = data_models.PolytopeSampler(domain=domain)
sampler = strategies.PolytopeSampler(data_model=data_model)

# this generates the samples, it is currently a bit slow as we generate the complete combinatorics, this can be easily speedup
samples = sampler.ask(5, return_all=False)

# this gives you callables for nonlinear inequality constraint to handle it in optimize_acqf
callables = get_nchoosek_constraints(domain=domain)

```
This are the generated samples, fulfilling both constraints.

<img width="780" alt="image" src="https://user-images.githubusercontent.com/28678090/226090540-afc320c5-70e9-444c-84f1-2f71dd1c56cf.png">


In the next weeks this should be all automatically get integrated. Note that bofire is still in a kind of alpha stage ;)

**Comment by sgbaird** [comment]
@jduerholt thanks for sharing [BoFire](https://github.com/experimental-design/bofire)! Excited to see where this goes. For the discussions you mentioned:
- https://github.com/facebook/Ax/issues/769
  - Tutorial notebook https://github.com/facebook/Ax/issues/769#issuecomment-1021471675
- https://github.com/facebook/Ax/issues/727

**Comment by Balandat** [comment]
Glad to hear you were able to make some progress on this. 

> We did run into the next problem, which is that nonlinear constraints have not been implemented in botorch for q>1 (or for knowledge gradient). These constraints are handled through the [make_scipy_nonlinear_inequality_constraints](https://github.com/pytorch/botorch/blob/main/botorch/optim/parameter_constraints.py#L414) function, which works differently from the linear one.

Doing this should hopefully not be too hard if the constraints are intra-point constraints that apply separately to each of the `q` candidates (rather than a constraint across the different candidates in a batch). Looks like that is your use case?

> Out of 10 possible features, we can only select 5 at a time for any sample in the q-batch.

For KG it's going to be a bit trickier, since the locations of the fantasy observations also need to satisfy those constraints. That won't really be an overly challenging problem to hook up, but my concern is that you'll end up with a huge number of nonlinear constraints in the acquisition function optimization problem that way. By default that optimization will use SLSQP which is exceedingly slow with many constraints. Do you need to use KG here?

**Comment by swierh** [comment]
> Doing this should hopefully not be too hard if the constraints are intra-point constraints that apply separately to each of the `q` candidates (rather than a constraint across the different candidates in a batch). Looks like that is your use case?

Yes, the points in a batch are independent. So far I've given it one attempt, but got stuck on the shape of [`x0`](https://github.com/pytorch/botorch/blob/main/botorch/generation/gen.py#L225) causing a bunch of conflicts. I'll hopefully have some time soon to give it another go.

> For KG it's going to be a bit trickier, since the locations of the fantasy observations also need to satisfy those constraints. That won't really be an overly challenging problem to hook up, but my concern is that you'll end up with a huge number of nonlinear constraints in the acquisition function optimization problem that way. By default that optimization will use SLSQP which is exceedingly slow with many constraints. Do you need to use KG here?

We can use other acquisition functions, though KG seems to perform well on our specific challenge. As for it being exceedingly slow, for our challenge it's not really a problem if the computation of a single batch takes a few hours, so that might be something we can test.

**Comment by jduerholt** [comment]
> Yes, the points in a batch are independent. So far I've given it one attempt, but got stuck on the shape of [`x0`](https://github.com/pytorch/botorch/blob/main/botorch/generation/gen.py#L225) causing a bunch of conflicts. I'll hopefully have some time soon to give it another go.

@swierh: are you already on it making the non-linear constraints also possible for q>1? Else, I can also give it a try as we also need it. I tested so far only with q=1. A quick fix could also be to optimize with `sequential = True`.

@Balandat: concerning speed, for larger q-batches and higher dimensional problems with a lot of constraints, one could also try how `ipopt` peforms instead of SLSQP. Could be faster there, and the interface is the same as for the scipy optimizers.

**Comment by Balandat** [comment]
> one could also try how ipopt peforms instead of SLSQP

Yeah that's also possible. I've used IPOPT quite a bit in grad school, at the time it wasn't exactly easy to build and run on different platforms with a python interface - but I assume this has improved since? If you gave this a try I'd love to hear about how it goes.

**Comment by jduerholt** [comment]
We are using it to generate D-optimal designs using `cyipopt` as python wrapper. Installation is still not so easy, especially with pip. As soon as I have time, I will give it a try for acqf optimization ;)

**Comment by swierh** [comment]
@jduerholt I'm not currently working on it, and won't have much time for at least the coming week, so feel free to give it a try.

**Comment by jduerholt** [comment]
@swierh: the PR regarding problems with nonlinear constraints with q > 1 has been merged last week (https://github.com/pytorch/botorch/pull/1793), and the feature is now available in the main branch.

**Comment by swierh** [comment]
@jduerholt Awesome! Thanks so much for the effort you and @Balandat put into it!

Resolved by https://github.com/pytorch/botorch/pull/1793

**Comment by sgbaird** [comment]
Noting that a new NChooseK scheme was [recently implemented in BoFire](https://github.com/experimental-design/bofire/releases/tag/v0.4.0), based on the recent article, "BONSAI" https://arxiv.org/abs/2602.07144

xref:
- https://github.com/experimental-design/bofire/pull/747
- https://github.com/experimental-design/bofire/pull/752
